enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,318 @@
+---
+title: configurable-dns-forwarding
+authors:
+  - "@dhansen"
+reviewers:
+  - "@ironcladlou"
+  - "@Miciah"
+  - "@frobware"
+approvers:
+  - "@knobunc"
+  - "@ironcladlou"
+creation-date: 2019-09-27
+last-updated: 2019-10-01
+status: implementable
+see-also: 
+replaces:
+superseded-by:
+---
+
+# Configurable DNS Forwarding
+
+This proposal allows cluster operators the ability to configure DNS name query forwarding.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+1. Is `operator.openshift.io` the best API group for the proposed `Upstream` type?
+2. How are API changes handled for cluster downgrades?
+3. Should IPv6 references for `Nameservers` be removed until OpenShift adds IPv6 support?
+4. Is the cluster domain (i.e. cluster.local) an invalid `Upstream` domain?
+
+## Summary
+
+Currently, the [cluster-dns-operator](https://github.com/openshift/cluster-dns-operator) manages DNS forwarding with a
+static ConfigMap:
+
+```bash
+$ oc get configmap/dns-default -n openshift-dns -o yaml
+apiVersion: v1
+data:
+  Corefile: |
+    .:5353 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+            pods insecure
+            upstream
+            fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        forward . /etc/resolv.conf
+        cache 30
+        reload
+    }
+kind: ConfigMap
+metadata:
+  labels:
+    dns.operator.openshift.io/owning-dns: default
+  name: dns-default
+  namespace: openshift-dns
+  <SNIP>
+```
+
+This proposal will generate this ConfigMap with a `forward` section based on user-provided values of type `Upstream`.
+
+## Motivation
+
+DNS name resolution for pods is currently provided by an instance of [CoreDNS](https://coredns.io) that runs on each
+node in the cluster. If a DNS query is for a hostname within the cluster domain suffix (i.e. `cluster.local`), CoreDNS
+resolves the name. Otherwise, CoreDNS proxies DNS messages to an upstream resolver identified by `/etc/resolv.conf` on
+the corresponding node. Although this provides a consistent and reliable approach for name resolution, it restricts how
+cluster operators manage DNS name queries.
+
+### Goals
+
+1. A well defined API for DNS forwarding configuration that is implementation agnostic.
```

Comment:
What other non-CoreDNS implementations could theoretically support the proposed API?

Reply:
@danehans @smarterclayton 

I've been thinking about how quickly the API proposed here has expanded to include domain-mapping and forwarding policies. We started with `[]string` (which I still _think_ could have been made to work for the real world use cases I know of, although I'm open to being wrong about that.)

Is it fair to go even further and ask what other Kube-compliant DNS servers exist now or on the horizon which could possibly replace CoreDNS even for minimal functionality (Kube DNS compliance), let alone one that may in addition be easily made to support the API we're describing?

I ask these questions because if this is _really_ just a CoreDNS API, there could be a discussion about the pros and cons or trying to abstract that fact at all.

Reply:
> What other non-CoreDNS implementations could theoretically support the proposed API?

[kube-dns](https://kubernetes.io/blog/2017/04/configuring-private-dns-zones-upstream-nameservers-kubernetes/)

Reply:
Isn't that SkyDNS? Doesn't CoreDNS replace it?

Reply:
@ironcladlou Although kube-dns and PowerDNS are theoretical options, I believe CoreDNS will  continue to be the de facto standard.

Reply:
CoreDNS is the default, but [kube-dns](https://kubernetes.io/docs/tasks/administer-cluster/dns-custom-nameservers/#kube-dns) can still be used and is supported upstream.

Reply:
I see no scenario where we switch off coredns or support another one - there is no use case for it.

Reply:
To put it another way, I see no reason to constrain to lowest common denominator for DNS config when we plan to only ship one dns provider for the next 1-2 years

Reply:
What are some concrete ways to leverage that assumption?

For example, it seems we're building API that matches the current plugin object and copying relevant docs from upstream to the API. So there's a design advantage there (we know the upstream object is useful and relatively stable). But are there additional risks to consider, and is there anything we should do to minimize them?

What else does coupling to CoreDNS let us do that we aren't already covering in the proposal?

Maybe the answer is "nothing," I'm just trying to figure out if there's anything I'm missing.

Reply:
> What else does coupling to CoreDNS let us do that we aren't already covering in the proposal?

Making `Domain` a list would hew more closely to the CoreDNS API.

Reply:
And I'd rename `Upstream` to something more generic (what we're calling an "upstream" translates to a CoreDNS server block).

Reply:
> Making Domain a list would hew more closely to the CoreDNS API.

@Miciah can you help me better understand your perspective? Each forward plugin consists of a `FROM`, consisting of the base domain to match for the request to be forwarded.

I have updated the `Upstream` type to `Forwarder` to better align with the CoreDNS naming scheme.

Reply:
> Each forward plugin consists of a `FROM`, consisting of the base domain to match for the request to be forwarded.

That will be `.`, right?  So input like the following: